### ISSUE TITLE
Proposal Filtering Implementation 

### ISSUE THREAD

**Issue by monicasarbu** [elastic/libbeat#336]
Alternative names: conditionals

The goal here is to reduce the number of exported events and fields, before they are sent over the network. It is not to add or change fields, that can be achieved later by Logstash. 

__Use cases to solve__

 - Filebeat: Ship only log lines that don’t match a certain regexp: https://github.com/elastic/filebeat/issues/78
 - Packetbeat: Drop all 200 OK transactions (not yet requested)
 - Topbeat: Export all metrics with higher cpu usage (not yet requested)

__Requirements__

- define a generic filtering that is implemented in libbeat and used by all the Beats, instead of having a specific filtering implementation for each Beat. 
- be able to use event fields (e.g. http.response.code) in the condition
- be able to drop the event or remove a certain field if the condition is fulfilled 

__Disadvantages__

If we want to use generic conditions in the filtering, then we need to apply the filtering on the created event object. In other words, we need to build the event object before running the filtering condition. Building the event object might be expensive as we need to calculate all the fields and if we just decide to drop the event, then the time spent is useless.

To improve this, I propose to decide about a minimum set of fields to be exported by each Beat. Additional fields can be added by enabling options in configuration file. These options will have to be implemented in each Beat but have the advantage of having the maximum performance. For example, Filebeat can implement an option like “exclude_files” a lot more efficient than it can be done in the generic filtering.

__Proposal__

The filters are configured as a YAML list, each with an _if_ and _do_ blocks. The _if_ block is optional. The conditions are modeled roughly after the Elasticsearch filtering. Each action can receive parameters.

The filtering rules are executed in libbeat before publishing the event.

<pre>
filters:
    - if:
        condition
      do:
        - action1
        - action2

    - if:
        condition
      do
        - action1
        - action2
        - action3

</pre>

__Supported actions:__

- drop_event (phase 1)
- drop_fields (phase 2)

__Conditions:__

- equals
- contains
- regexp
- range
- and (phase 2)
- or (phase 2)
- not (phase 2)

__Examples__

- Filebeat drops lines starting with DBG

<pre>
filters:
    - if:
        regexp:
            message: “^DBG” 
      do:
          - drop_event: true
</pre>

- Filebeat drops line starting with DBG for apache logs only

<pre>
filters:
    - if:
        and:
            equals:
                type: apache
            regexp:
                message: “^DBG”
      do:
          - drop_event: true

</pre>

in prospector:
<pre>
		document_type: apache
</pre>
	
- Drop all 200 OKs in Packetbeat

<pre>

filters:
      - if:
          equals: 
	        “http.response.code”: 200
       do:
           - drop_event: true

</pre>

**Comment by ruflin** [comment]
This looks like a quite nice general purpose solution. I can see this as the end result where we end up with the full implementation.

I would suggest that we start small with a minimal implementation in filebeat and learn how it works and people use it and later generalise it into libbeat. For the first implementation "regexp" and "drop_event" seems to be sufficient to me.

For the YAML config options: On the one hand I quite like it as it builds up on an existing configuration language without modifying it but at the same time it feels a little bit unnatural. Is this based on a common way how if / else is done in YAML? 

An alternative to the above could be to use a notation very similar to the query_string notation: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html This would allow to use a "common" format and not have the full YAML config options. 

I think we have to be careful about the drop_field part as this can also break dashboards. At the same time people that use it should know what they are doing.

**Comment by kimchy** [comment]
I like the suggestion, my comment is that we should aim to make the most common use case as concise as possible. If we think that dropping events would be the most common one, then the suggested syntax is a bit verbose? 

I don't have a good solution, but conceptually, @ruflin is an interesting one where we can have a mini conditional language here, that is generic, and then flip the action with the condition, something like this:

```
drop_event: message~=^DBG && type=apache
```

we can then use this conditional syntax and code element in several other places, for example, "drop_field", or "sampling" and so on.

While I write this, even if we go with more structured language, feels to me that it is better to flip the action and the condition, so we can use it as an element in other places code wise, so something like this:

```
drop_event:
      equals:
            type: apache
```

```
drop_event:
        and:
            equals:
                type: apache
            regexp:
                message: “^DBG”
```

Then we can also do (in the future):

```
sample:
     when:
         equals:
               type: apache
     interval: 1s
     ....
```

And the conditional language remains the same.

we could eventually evolve to both, similar to how in ES you can do both structured and unstructured, it is just a "conditional DSL", and we can start with the structured one to begin with, and slowly evolve?

**Comment by ruflin** [comment]
Perhaps we should also take the multiline "conditionals" into account here. Based on the content of a log line it is decided what the actions are: https://github.com/elastic/filebeat/issues/301 This is almost identical to the filtering, just that actions based on the results are different. But at the same time, it only applies to one field. Not sure if adds more complexity mixing these two instead of simplifying it.

**Comment by monicasarbu** [comment]
@kimchy  Initially I thought of something very similar with your proposal, that is able to execute a single action per condition, but then I realized that it might happen that you need to execute multiple actions for the same condition. This way I came up with the if - do solution, that is more verbose. For example for 2xx responses you might want to reduce the number of fields and also sample at a higher rate. To be honest, at the moment, I am not sure if this will be common enough, so I agree with your proposal.  

I like the idea of starting with structured syntax and then add the unstructured one.

Just to clarify, the syntax for dropping fields could be something like this:

<pre>
- drop_fields:
  fields: ['load', 'swap']
  when:
    type: system
</pre>

And the syntax for dropping an event:

<pre>
- drop_event:
  equals: 
    “http.response.code”: 200
</pre>

**Comment by monicasarbu** [comment]
The issue is copied in https://github.com/elastic/beats/issues/451 and becomes a duplicate.

**Comment by urso** [comment]
I like the samples given last by @kimchy and @monicasarbu. Maybe we can make the
predicate sublanguage mostly similar to the lucene query language (later).

Using the filters as described, it seems that we have to test/apply every single
filter if multiple filters are given. In case we want to share a condition with
multiple filter operators, we'd have to switch back to the original proposal.
Still it's possible to combine the original proposal with conditions in filter operators.

e.g.

    filters:
      - if:
          cond:
            regexp:
              message: "^DBG"
          actions:
            - drop_event: # maybe introduce select/filter instead of drop_event
                when:
                  equals:
                    type: apache
            - drop_fields:
                fields: ['load', 'swap']
                when:
                  type: system


We should also stick to single key operators. Like this 'if' having multiple options 'cond' for the condition and 'actions' for the body to execute.
  
Is there any interest in configurable chaining of operations?

I've been thinking of having something similar to pipelined aggregations for
filtering. (which kinda resembles a stream process operators). Basically a small
stream processing language employed by the publisher in libbeat for use by
administrators. Every event being forwarded to libbeat will be processed by it.


That is I'm thinking about having basically 3 languages build on top of each other:
1. predicate language (reusable by different components, e.g. filebeat line filtering)
2. filter operator language (e.g. drop_event, select, drop_fields), reusing predicate language for conditionals.
3. streaming language with configurable data flow, reusing filter operators and predicate language.


Some potential use cases:

1. reduced reporting period (like rate limiting, but based on process name and low cpu usage) for processes with low resource usage:

        # filter all input events of type 'process' into 'processes' stream
        processes:
          select:  # select similar to filter or negated drop_events
            where:
              type: process
    
        # filter all input events of type 'load' into 'load' stream
        sysload:
          select:
            where:
              type: load
    
        cpu_events:  # name of event stream
          - split:
              path: processes
              cond:
                - low: # name of sub-event stream (access via cpu_events.low)
                  where:
                    less:
                    cpu: 60
                - high: # name of sub-event stream (access via cpu_events.high)
                  otherwise:
    
        low_resource_usage: # name of rate limited event stream
          periodic:
            path: cpu_event.low
            period: 60s
            key: proc.name
    
        # output stream with all events to be published and removed fields
        output:  
          merge: # merge multiple event streams into 1 stream
            - sysload
            - low_resource_usage
            - cpu_events.high

2. drop http events with 200 response + remove some fields

        output:
          # anonymous chaining (without named stream/pipeline) generates
          # straight pipeline
          - drop_event:
              equals: 
                “http.response.code”: 200
    
          - drop_fields:
              fields: ['status']

3. correlation of http + mongodb transactions based on regex and timed window

        all: # all input events published to libbeat (due to no filter/operator)
    
        http: # stream of windowed batches of http transactions
          - select:
              where:
                type: http
          - window
              duration: 5s
              op: sliding
    
        mongodb: # stream of windowed batches of mongodb transactions
          - select:
            where:
              type: mongodb
    
        correlated: # stream of correlation events
          join:
            paths:
              h: http
              m: mongodb
            where:
              regex:
                pattern_from: h.http.body.user-id # via JSON body parser
                in: "m.body"
            fields: # fields to be reported in new event
              ...
    
        output:
          merge:
            - all
            - correlated


Some common operators found in stream processing languages:

- projection: add, remove, update fields
- selection/filtering: select or remove events based on predicate
- aggregation: statefull operator on windows (batch of events)
- join: correlate event from multiple windows
- split: divide input stream into multiple streams based on condition (conditional fan-out)
- barrier: synchronize streams
- merge: (fan-in) combine multiple streams into one output stream
- window: combine multiple events into batches. Windows can be time and/or size based + sliding or non-overlapping

**Comment by ruflin** [comment]
@monicasarbu Not sure if it makes sense to move such issues in the middle of the discussion as it makes it hard to follow the comments. As I was confused first, I made it now more obvious in the other issue that there was some discussion / comments here.

**Comment by ruflin** [comment]
@urso Your proposal is quite interesting and I like it. What I worry is that we add too much complexity on the client (beats) side and beats get more resource hungry then they should. We should find a good balance between not sending all data to reduce network traffic and simple filtering.
